test/unit/puppetlabs/puppetserver/ruby/http_client_test.clj

```diff
@@ -76,82 +76,120 @@ jruby-config :- jruby-schemas/JRubyConfig
-                                                     port use-ssl?)))
-     (doto scripting-container
-       (.runScriptlet (format "Puppet[:hostcert] = '%s'" cert-pem))
-       (.runScriptlet (format "Puppet[:hostprivkey] = '%s'" privkey-pem))
-       (.runScriptlet (format "Puppet[:localcacert] = '%s'" ca-pem)))
-     scripting-container)))
+  []
+  (let [jruby-config (jruby-config)
+        scripting-container (-> (jruby-impl-core/create-pool jruby-config)
+                                (jruby-core/borrow-from-pool :http-client-test [])
+                                :scripting-container)]
+    (doto scripting-container,
+      (.runScriptlet "require 'puppet'")
+      (.runScriptlet (format "Puppet[:hostcert] = '%s'" cert-pem))
+      (.runScriptlet (format "Puppet[:hostprivkey] = '%s'" privkey-pem))
+      (.runScriptlet (format "Puppet[:localcacert] = '%s'" ca-pem)))
+    scripting-container))
+
+(defn create-http-client
+  [scripting-container port options]
+  (.runScriptlet scripting-container "require 'puppet/server/http_client'")
+  (let [use-ssl? (true? (:use-ssl options))
+        http-client-settings (get-http-client-settings
+                              (select-keys options [:ssl-protocols
+                                                    :cipher-suites]))
+        http-client-class (.runScriptlet scripting-container
+                                         "Puppet::Server::HttpClient")]
+    (.callMethod scripting-container http-client-class
+                 "initialize_settings" http-client-settings Object)
+    (.runScriptlet scripting-container
+                   (format "$c = Puppet::Server::HttpClient.new('localhost', %d, {:use_ssl => %s});"
+                           port use-ssl?))))
+
+(defn terminate-http-client
+  [scripting-container]
+  (.runScriptlet scripting-container "$c = nil")
+  ;; This step purges the '@ssl_context' which may have been used with an http
+  ;; client call.  This is necessary to ensure that any subsequent http client
+  ;; call utilizes a new SSL context rather than unintentionally using some
+  ;; cached state - e.g., ssl parameters from prior connection attempt.
+  (.runScriptlet scripting-container
+                 "Puppet::Server::Config.instance_variable_set('@ssl_context', nil)")
+  (.runScriptlet scripting-container "Puppet::Server::HttpClient.terminate"))
 
 (defn terminate-scripting-container
   [scripting-container]
-  (.runScriptlet scripting-container "Puppet::Server::HttpClient.terminate")
   (.terminate scripting-container))
 
+(def ^:dynamic ^:private *scripting-container* nil)
+
+(defmacro with-scripting-container
+  [scripting-container & body]
+  `(let [~scripting-container (or *scripting-container*
+                                  (create-scripting-container))]
+     (try
+       ~@body
+       (finally
+         (when-not *scripting-container*
+           (terminate-scripting-container ~scripting-container))))))
```

Comment:
Hmm.  You could still terminate it if it was memoized but I guess it wouldn't be freed because there'd still be a reference to it.

meh.  i guess it's fine this way.